src/CrossChainReceiverFactory.sol

```diff
@@ -0,0 +1,406 @@
+                    revert(ptr, returndatasize())
+                }
+            }
+        }
+        assembly ("memory-safe") {
+            let ptr := mload(0x40)
+
+            mstore(0x00, 0x095ea7b3) // selector for `approve(address,uint256)`
+            mstore(0x20, 0x000000000022D473030F116dDEE9F6B43aC78BA3) // Permit2
+            mstore(0x40, amount)
+
+            if iszero(call(gas(), token, 0x00, 0x1c, 0x44, 0x00, 0x20)) {
+                returndatacopy(ptr, 0x00, returndatasize())
+                revert(ptr, returndatasize())
+            }
+            if iszero(or(and(eq(mload(0x00), 0x01), lt(0x1f, returndatasize())), iszero(returndatasize()))) {
+                mstore(0x00, 0x3e3f8f73) // selector for `ApproveFailed()`
+                revert(0x1c, 0x04)
+            }
+
+            mstore(0x00, 0x01)
+            return(0x00, 0x20)
+        }
+    }
+
+    function call(address payable target, uint256 value, bytes calldata data)
+        external
+        onlyOwner
+        returns (bytes memory)
+    {
+        assembly ("memory-safe") {
+            let ptr := mload(0x40)
+
+            calldatacopy(ptr, data.offset, data.length)
+            let success := call(gas(), target, value, ptr, data.length, 0x00, 0x00)
+
+            returndatacopy(add(0x40, ptr), 0x00, returndatasize())
+
+            if iszero(success) { revert(add(0x40, ptr), returndatasize()) }
+
+            mstore(add(0x20, ptr), returndatasize())
+            mstore(ptr, 0x20)
+            return(ptr, add(0x40, returndatasize()))
+        }
+    }
+
+    function cleanup(address payable beneficiary) external {
+        if (msg.sender == address(_cachedThis)) {
+            selfdestruct(beneficiary);
+        }
+
+        address owner_ = owner();
+        if (_msgSender() != owner_) {
+            if (owner_ != address(0)) {
+                _permissionDenied();
+            }
+            address pendingOwner_ = pendingOwner();
+            if ((pendingOwner_ == address(0)).or(beneficiary != pendingOwner_)) {
+                _permissionDenied();
+            }
+        }
+        selfdestruct(beneficiary);
+    }
+
+    function _erc1271IsValidSignature(bytes32 hash, bytes calldata signature, address owner)
+        internal
+        view
+        virtual
+        returns (bool result)
+    {
+        bytes32 nameHash = _NAMEHASH;
+        assembly ("memory-safe") {
+            let ptr := mload(0x40) // Grab the free memory pointer.
+            // Skip 2 words for the `typedDataSignTypehash` and `contents` struct hash.
+            mstore(add(0x40, ptr), nameHash)
+            mstore(add(0x60, ptr), chainid())
+            mstore(add(0x80, ptr), address())
+            
+            // `c` is `contentsDescription.length`, which is stored in the last 2 bytes of the signature.
+            let c := shr(0xf0, calldataload(add(signature.offset, sub(signature.length, 0x02))))
+            for {} 1 {} {
```

Comment:
I don't understand the heuristic applied to this scoping. Why are some things inside the `for` and others outside?

Reply:
this trick with the `for` is typically used because it coaxes solc into generating more efficient code for the branching. it's preferred anytime there's control flow that's more complex than a simple `if` without `else`. it's worth testing, though, now that we've significantly altered the original code

Reply:
Ended up removing the `for` anyway

Reply:
can you double check that this actually improves gas usage? the `for` trick tends to be the most optimal

Reply:
> solc really does crazy stuff 😅

unfortunately, this is the case :crying_cat_face: it's always best to check by running the test and noting the gas

Reply:
That's a huge difference! I am very suspicious

Reply:
you were right on being suspicious, just double checked, chaining it to the `for`. It saves a bit of gas, kept as much code outside of the `for` as posible because it was gives the best gas results.